copy.lua

```diff
@@ -0,0 +1,417 @@
+            self.offsetDirection=(self.offsetDirection+1)%4
+        elseif keys.CUSTOM_T then
+            self.buffer=self:transformBuffer(function(x,y,xlen,ylen,tile) return ylen-y,x  end)
+            self.offsetDirection=(self.offsetDirection-1)%4
+        elseif keys.CUSTOM_G then
+            self.offsetDirection=(self.offsetDirection+1)%4
+        elseif keys.CUSTOM_E then
+            self.option=self.option=='erase' and 'normal' or 'erase'
+        elseif keys.CUSTOM_X then
+            self.option=self.option=='construction' and 'normal' or 'construction'
+        elseif keys.CUSTOM_I then
+            self:invertBuffer()
+        elseif keys.CUSTOM_C then
+            self.state='convert'
+        elseif keys.CUSTOM_B then
+            self.blink = not self.blink
+        elseif keys.SELECT then
+            self:pasteBuffer(copyall(df.global.cursor))
+        end
+    elseif self.state=='mark' then
+        if keys.SELECT then
+            if self.marking then
+                --set the table
+                self.state='paste'
+                self.marking = false
+                self:setBuffer(getTiles(self.mark,copyall(df.global.cursor),self.cull))
+            else
+                self.marking = true
+                self.mark = copyall(df.global.cursor)
+            end
+        elseif keys.LEAVESCREEN then
+                self.state='paste'
+                return
+        elseif keys.CUSTOM_P then
+            self.cull = not self.cull
+        end
+    elseif self.state=='convert' then
+        if keys.LEAVESCREEN then
+            self.state='paste'
+            return
+        elseif keys.CUSTOM_D then
+            self:transformBuffer(function(x,y,xlen,ylen,tile) if tile.dig>0 then tile.dig=1 end return x,y end)
+            self.state='paste'
+        elseif keys.CUSTOM_H then
+            self:transformBuffer(function(x,y,xlen,ylen,tile) if tile.dig>0 then tile.dig=3 end  return x,y end)
+            self.state='paste'
+        elseif keys.CUSTOM_U then
+            self:transformBuffer(function(x,y,xlen,ylen,tile) if tile.dig>0 then tile.dig=6 end  return x,y end)
+            self.state='paste'
+        elseif keys.CUSTOM_J then
+            self:transformBuffer(function(x,y,xlen,ylen,tile) if tile.dig>0 then tile.dig=5 end  return x,y end)
+            self.state='paste'
+        elseif keys.CUSTOM_I then
+            self:transformBuffer(function(x,y,xlen,ylen,tile) if tile.dig>0 then tile.dig=2 end  return x,y end)
+            self.state='paste'
+        elseif keys.CUSTOM_R then
+            self:transformBuffer(function(x,y,xlen,ylen,tile) if tile.dig>0 then tile.dig=4 end  return x,y end)
+            self.state='paste'
+        end
+        
+    end
+    if keys.LEAVESCREEN then
+            self:dismiss()
+    elseif self:propagateMoveKeys(keys) then
+        return
+    end
+end
+
+--------------------------WARNING: JANKY SHIT BELOW--------------------------
+for i=1, 4 do
```

Comment:
I think what gui/teleport does is something you could try, and that could be moved inside the CopyUI class too.

https://github.com/DFHack/scripts/blob/0b6e4a56392e125c01e24694ddd880bbc652451d/gui/teleport.lua#L47-L63

Forcing ui.main.mode to `df.ui_sidebar_mode.LookAround` might be more appropriate for your case. One benefit of that is that it will make DF show the sidebar *temporarily* if it was hidden before - your way overrides user preferences.

~~You probably also don't need to copy/restore the sidebar mode like that script - that was an old feature for supporting the `k` (LookAround) menu too, but I don't think it works any more.~~ Looks like you're already doing this.